### PR TITLE
API v0 and admin interface

### PR DESCRIPTION
**Pull request by pirtnea**
Implemented API v0 for regular users and admin interface for administrators.
Broke tests, fixed tests.
 :sparkles: :trumpet:
:notes:  :sparkles:

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,48 +1,21 @@
...
-  rescue_from ApplicationController::NoExerciseError do
-    render_error_page(status: 400, text: 'No exercises for this assignment')
-  end
-
-  def ensure_signed_in!
-    return if current_user
-
-    raise NotLoggedIn
+  rescue_from NotAuthorized do
+    redirect_to sessions_path, alert: 'Not authorized.'
```
Comment (vilmakoo) [inline app/controllers/application_controller.rb]: Why period here but not in most of the error messages? Feels inconsistent. :shipit:

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -51,16 +24,16 @@ def admin?
...
-    raise ApplicationController::NotAuthorized unless admin?
+    raise NotAuthorized unless admin?
   end
 
-  def only_user_and_admins!(user)
-    raise ApplicationController::NotAuthorized unless admin? || current_user == user
+  def ensure_signed_in!
+    return if current_user
+    session[:return_to] = request.url
+    redirect_to sessions_path, notice: 'Please log in.'
```
Comment (vilmakoo) [inline app/controllers/application_controller.rb]: :black_circle:

File: app/views/exercises/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    <%= form.label :model_solution %>
+    <%= form.text_area :model_solution, id: :exercise_model_solution %>
+  </div>
+
+  <div class="field">
+    <%= form.label :template %>
+    <%= form.text_area :template, id: :exercise_template %>
+  </div>
+
+    <p>Please, don't do this to me.</p>
```
Comment (vilmakoo) [inline app/views/exercises/_form.html.erb]: Who is "me"?  :confused:

File: config/routes.rb [hunk 4]
```diff
@@ -11,10 +28,17 @@
   resources :exercises
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
 
-  post 'exercises/:id/results', to: 'exercises#sandbox_results'
-
-  get 'peer_reviews/assignments/:assignment_id/request_exercise', to: 'peer_reviews#assign_exercise'
-
-  get 'peer_reviews/exercises/:id/stub_zip', to: 'peer_reviews#send_stub_zip'
-  get 'peer_reviews/exercises/:id/model_zip', to: 'peer_reviews#send_model_zip'
+  namespace :api do
```
Comment (vilmakoo) [inline config/routes.rb]: :scream_cat:
What these rows do??

File: spec/controllers/peer_reviews_controller_spec.rb [hunk 5]
```diff
@@ -1,90 +0,0 @@
-# frozen_string_literal: true
-
-require 'rails_helper'
-
-RSpec.describe PeerReviewsController, type: :controller do
```
Comment (vilmakoo) [inline spec/controllers/peer_reviews_controller_spec.rb]: :wave:

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (pirtnea) [inline app/controllers/application_controller.rb]: Because Henrik.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (pirtnea) [inline app/controllers/application_controller.rb]: :white_circle:

File: app/views/exercises/_form.html.erb (hunk 3 above)
Comment (pirtnea) [inline app/views/exercises/_form.html.erb]: Crowdsorceress  :dancer: :sparkles:

File: config/routes.rb (hunk 4 above)
Comment (pirtnea) [inline config/routes.rb]: [Pure magic!](http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/)

File: spec/controllers/peer_reviews_controller_spec.rb (hunk 5 above)
Comment (pirtnea) [inline spec/controllers/peer_reviews_controller_spec.rb]: :raised_hands:

File: Gemfile [hunk 6]
```diff
@@ -61,3 +61,9 @@ gem 'netrc', '~> 0.11.0'
 gem 'rest-client', '~> 1.8'
 
 gem 'paper_trail'
+
+gem 'bootstrap', '~> 4.0.0.beta'
+gem 'jquery-rails'
```
Comment (nygrenh) [inline Gemfile]: Please lock this version.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (nygrenh) [inline app/controllers/application_controller.rb]: Yes

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (nygrenh) [inline app/controllers/application_controller.rb]: :balloon:

File: config/routes.rb [hunk 7]
```diff
@@ -1,6 +1,23 @@
 # frozen_string_literal: true
 
+module ActionDispatch
+  module Routing
+    class Mapper
+      def resources3(*args)
```
Comment (nygrenh) [inline config/routes.rb]: Good code :+1:

Comment (vilmakoo) [inline config/routes.rb]: No

File: app/controllers/api/v0/exercises_controller.rb [hunk 8]
```diff
@@ -11,7 +11,7 @@ def create
         @exercise.attributes = exercise_params
 
         if @exercise.in_progress?
-          render json: { message: 'Exercise is already in progress', exercise: @exercise, status: 400 }
+          render json: { message: 'Exercise is already in progress.', exercise: @exercise, status: 400 }
```
Comment (vilmakoo) [inline app/controllers/api/v0/exercises_controller.rb]: Good code. :smiley:

Comment (pirtnea) [inline app/controllers/api/v0/exercises_controller.rb]: Yes

File: Gemfile.lock [hunk 9]
```diff
@@ -220,32 +220,32 @@ PLATFORMS
...
-  factory_girl_rails
+  byebug (~> 9.0.6)
+  dotenv-rails (~> 2.2.1)
+  factory_girl_rails (~> 4.8.0)
   http-cookie (~> 1.0, >= 1.0.3)
-  jquery-rails
+  jquery-rails (~> 4.3.1)
   listen (>= 3.0.5, < 3.2)
-  mime-types
+  mime-types (~> 2.99.3)
```
Comment (vilmakoo) [inline Gemfile.lock]: Good fix :+1:

**Pull request merged**